### DISCUSSION TITLE
Cannot see crossplane provisioned kubernetes objects in argocd UI

### DISCUSSION THREAD

**Discussion by MansiVerma777** [crossplane/crossplane#4704] [Q&A] [answered]
Hi,

We are currently running some evaulation around crossplane ARGOCD integration for kubernetes deployements.

We use crossplane kubernetes provider to create managed resources which are deployments/service/ingress objects on the k8 cluster.

Consider this scenario-
1) Devops exposed XRDs for app teams for easy deployment
2) User commits app claims to git repo
3) ARGOCD application that looks at this git repo deploys resources to cluster

Now the issue here is we are just able to see XRDs and MRs on argocd UI, we are unable to visualise the deployments/service/ingress objects that were deployed on the cluster in argocd UI. So we are evaulation a workaround -

1) Devops exposed XRDs for app teams for easy deployment
2) Devops evaluate app team's XRDs to generate MR definitions through crossplane k8 provider. Crossplane should not deploy anything here, just generate manifest and we use some crossplane provider to commit these definitions to git
3) ARGOCD application that looks at this git repo deploys resources to cluster

This way we see end resources in ARGCD UI
Is the workaround feasible? Or does community suggest some other way to achieve this?

**Comment by turkenh** [comment, marked as answer]
Not an ArgoCD expert but wondering if there are ways to configure ArgoCD to track those resource managed with provider-kubernetes with the help of additional labels/annotations? https://argo-cd.readthedocs.io/en/stable/user-guide/resource_tracking/

If there are, then you can add patch those to your provider-kubernetes `Object` and get them tracked by ArgoCD.

**Reply by mukul-vcc** [reply]
@turkenh - We are also using a similar pattern and in our case the objects are visible in the ArgoCD UI under the XR but we don't get to see the resources which are created by those objects. Do you have any recommendations for that? Wondering if the resources under objects should have had the ownerReferences pointing to the object automatically and if that would have solved this issue?

**Reply by bobh66** [reply]
@mukul-vcc it is not possible to add `ownerReferences` to the provider-kubernetes `Object` because the resource may be in a remote cluster and cross-cluster `ownerReferences` are not supported by kubernetes.

**Reply by mukul-vcc** [reply]
@bobh66 - What should be the best approach here to display the resources created by the object on the ArgoCD UI in a graceful manner? Just a note that in the current setup we are deploying the resources on the same cluster. The not so good solution that we have right now is to have `annotation+label` [resource tracking](https://argo-cd.readthedocs.io/en/stable/user-guide/resource_tracking/) for the resources created by the objects with `prune=false` [sync option](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#no-prune-resources) annotation applied on the resources to prevent deletion by ArgoCD. Finally [compare options](https://argo-cd.readthedocs.io/en/stable/user-guide/compare-options/) are enabled for excluding these resources from app's overall sync status. To sum it up, below configuration is used for being able to display the resources under the same app from which the composite resource is created. This keeps the overall status of the Argo app as synced but the individual resources still **seem** off with error as `OutOfSync (This resource is not present in the application's source. It will be deleted from Kubernetes if the prune option is enabled during sync.)`, which is expected in this scenario as the resources are not present in the application's source.

```
                    labels:
                      argocd.argoproj.io/instance: {{ $argoApp }}
                    annotations:
                      argocd.argoproj.io/tracking-id: {{ $resourceTrackingId }}
                      argocd.argoproj.io/sync-options: Prune=false
                      argocd.argoproj.io/compare-options: IgnoreExtraneous
```

**Reply by bobh66** [reply]
I'm afraid I don't know much about Grafana or Argo resource visualization.  There are a couple of projects that are focused on Crossplane-specific visualization - check out [komoplane](https://github.com/komodorio/komoplane) and [Crossview](https://github.com/corpobit/crossview) .

**Reply by mukul-vcc** [reply]
@bobh66 - I understand that, we are using ArgoCD primarily for the GitOps needs of our users and having a good visualisation for the kubernetes resources created by the objects of the kubernetes provider is a must have. It would have been really good to be able to see the resources created by the objects. We already use komoplane but for the users to go to different tools to see their resources is not something that we would want to have. Without being able to see the status of the kubernetes resources under the objects keeps the users wondering that the composite resource got created but still they don't get to perform the dependent tasks as the error from the resources are not visible to them.

**Reply by jonasz-lasut** [reply]
@mukul-vcc if you're using provider-kubernetes to create resources in the same cluster and same namespace as the XR then you can try to compose the resources directly from within the Crossplane composition if you're using v2. Then the resources have the ownerReferences set and ArgoCD can visualize the resources.

**Reply by mukul-vcc** [reply]
Hi @jonasz-lasut - We are not using v2 yet but will definitely try it once we move to v2.

**Reply by AdamS3x** [reply]
Hey all. So I'm also currently evaluating Crossplane v2 and while yes, using the direct composition does workaround this problem, it doesn't address the issues for those (such as myself) who want to use the more feature-rich kubernetes provider.

The main issue here is that the current version of the provider-kubernetes does not set the OwnerRef fields at all for objects which it creates i.e. the final Kubernetes objects such as Deployment/Service etc from the intermediate Object resource it itself is passed from the composition. It is these OwnerRef fields which K8s (and ArgoCD) use to identify parent/child relationships between resources, and is how most of the ArgoCD tree is built.

I've identified 2 ways to fix this visibility:

1)  As others are suggesting to patch in the ArgoCD tracking-id annotation:

  annotations:
    argocd.argoproj.io/tracking-id: >-
      argocd-app-name:xrd_api_version:namespace/resource-name

This is not ideal, as it makes any objects tagged with this appear as top-level resources which ArgoCD applied directly which is not the case as these are downstream resources created by Crossplane and should be visualised as such.

2) Fix the provider.
This is the only real solution for me, which is fix the provider-kubernetes to add the OwnerRef fields onto the objects it creates. So that is exactly what I've done - I've got a fixed version of the latest kubernetes-provider v1.2.1 which I've coded, and am testing locally, which has an opt-in approach to allow OwnerRef to be set on the child objects. This is enabled by setting a new annotation to true in the composition where you call the provider-kubernetes. The default behaviour is for this to be false if not set - for backwards compatibility with existing uses so it works exactly as it does now. I've not pushed this PR up yet, I'm in the process of having my changes code-reviewed internally but once that has been finished I will push it up as a PR to the project and if accepted will address everyone's issues as this is not the first time this limitation has been identified, so watch this space - I will post an update once this is ready.

**Comment by MoeidHeidari** [comment]
you can use this also https://corpobit.com/products/crossview

**Comment by bobh66** [comment]
@AdamS3x this is a known issue with the provider and unfortunately `ownerReferences` only work for objects in the local kubernetes cluster.  One of the major use cases of provider-kubernetes is to create resources in remote clusters, and since kubernetes does not support remote `ownerReferences` there is no easy way to solve that problem.  The provider has no way of knowing whether it is dealing with a local or remote resource.  See [this](https://github.com/crossplane-contrib/provider-kubernetes/pull/53) PR for more discussion.

**Reply by AdamS3x** [reply]
@bobh66 ah ok yes I see the problem. Is it still worth pushing up the PR or will this just cause issues for when the target is a remote cluster?

**Reply by bobh66** [reply]
If there was a way for the provider to know whether it is working on the local cluster it might be worth it, even though it would introduce an inconsistency between local and remote resources.  However if an invalid ownerReference is added to a remote resource (or a local resource for that matter) the resource will be deleted by kubernetes garbage collection, which is definitely not a desirable scenario.

Maybe a path forward would be to have an option that will automatically copy the ArgoCD tracking annotation - or all annotations? - from the Object to the wrapped resource, which would be safe to do for both local and remote resources.  It's not much of an improvement over doing the same thing in the Composition but it might be helpful.
